ipalib/rpc.py

```diff
@@ -274,102 +274,171 @@ def xml_dumps(params, version, methodname=None, methodresponse=False,
     )
 
 
-def json_encode_binary(val, version):
-    '''
-   JSON cannot encode binary values. We encode binary values in Python str
-   objects and text in Python unicode objects. In order to allow a binary
-   object to be passed through JSON we base64 encode it thus converting it to
-   text which JSON can transport. To assure we recognize the value is a base64
-   encoded representation of the original binary value and not confuse it with
-   other text we convert the binary value to a dict in this form:
-
-   {'__base64__' : base64_encoding_of_binary_value}
-
-   This modification of the original input value cannot be done "in place" as
-   one might first assume (e.g. replacing any binary items in a container
-   (e.g. list, tuple, dict) with the base64 dict because the container might be
-   an immutable object (i.e. a tuple). Therefore this function returns a copy
-   of any container objects it encounters with tuples replaced by lists. This
-   is O.K. because the JSON encoding will map both lists and tuples to JSON
-   arrays.
-   '''
-
-    if isinstance(val, dict):
-        new_dict = {}
-        for k, v in val.items():
-            new_dict[k] = json_encode_binary(v, version)
-        return new_dict
-    elif isinstance(val, (list, tuple)):
-        new_list = [json_encode_binary(v, version) for v in val]
-        return new_list
-    elif isinstance(val, bytes):
-        encoded = base64.b64encode(val)
-        if not six.PY2:
-            encoded = encoded.decode('ascii')
-        return {'__base64__': encoded}
-    elif isinstance(val, Decimal):
-        return unicode(val)
-    elif isinstance(val, DN):
-        return str(val)
-    elif isinstance(val, datetime.datetime):
-        if capabilities.client_has_capability(version, 'datetime_values'):
+class _JSONPrimer(dict):
+    """Fast JSON primer and pre-converter
+
+    Prepare a data structure for JSON serialization. In an ideal world, priming
+    could be handled by the default hook of json.dumps(). Unfortunately the
+    hook treats Python 2 str as text while FreeIPA considers str as bytes.
+
+    The primer uses a couple of tricks to archive maximum performance:
+
+    * O(1) type look instead of O(n) chain of costly isinstance() calls
+    * __missing__ and __mro__ with caching to handle subclasses
+    * inlined code with minor code duplication
+    * function default arguments to turn global into local lookups
+    * on-demand lookup of client capabilities with cache
+
+    Depending on the client version number, the primer converts:
+
+    * bytes -> {'__base64__': b64encode}
+    * datetime -> {'__datetime__': LDAP_GENERALIZED_TIME}
+    * DNSName -> {'__dns_name__': unicode}
+
+    The _ipa_obj_hook() functions unserializes the marked JSON objects to
+    bytes, datetime and DNSName.
+
+    :see: _ipa_obj_hook
+    """
+    __slots__ = ('version', '_cap_datetime', '_cap_dnsname')
+
+    _identity = object()
```

Comment:
Could you please add a comment explaining that `_identity = object()` with `if func is _identity then arg else func(arg)` is faster than `_identity = lambda x: x` with `func(arg)` so that no one comes in the future and replaces the former with the latter because it's duplicate code?